src/peft/tuners/lora/variants.py

```diff
@@ -130,6 +130,186 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
         return result
 
 
+class QALoraLinearVariant(LoraVariant):
+    @staticmethod
+    def init(module: Linear, adapter_name: str, **kwargs: Any) -> None:
+        """
+        Initializes QALoRA specific parameters for a given adapter.
+
+        Args:
+            module (Linear): The linear module to be adapted.
+            adapter_name (str): The name of the adapter.
+            **kwargs: Additional keyword arguments.
+                qalora_group_size (int, optional): The size of groups for pooling. Defaults to 16.
+        """
+        # Store the qalora_group_size as a module attribute
+        # It's expected that qalora_group_size is passed in kwargs from LoraConfig
+        qalora_group_size = kwargs.get("qalora_group_size", 16)  # Default to 16 if not specified in kwargs
+        if not hasattr(module, "qalora_group_size"):
+            module.qalora_group_size = {}
+        module.qalora_group_size[adapter_name] = qalora_group_size
+
+        # Store original dimensions for use in get_delta_weight
+        if not hasattr(module, "orig_in_features"):
+            module.orig_in_features = {}
+        module.orig_in_features[adapter_name] = module.in_features
+
+        # Create and store pooling factor for scaling
+        if not hasattr(module, "qalora_scaling_factor"):
+            module.qalora_scaling_factor = {}
+
+        if module.in_features % qalora_group_size == 0:
+            # Paper's pseudocode for QALoRA forward pass scales the pooled input x_pooled by (D_in / L_pseudo),
+            # where L_pseudo is the group_size.
+            # The get_delta_weight calculation also uses a paper_scaling_factor.
+            # For consistency with the paper's forward pass logic,
+            # this factor could represent (D_in / group_size) or simply group_size
+            # depending on interpretation.
+            # Let's assume paper_scaling_factor is related to how x_pooled is scaled before LoRA.
+            # The current implementation calculates it as module.in_features / qalora_group_size,
+            # which is the number of groups (L).
+            # If the intention is to scale by group_size, this should be `qalora_group_size`.
+            # For now, keeping the existing logic for qalora_scaling_factor:
+            module.qalora_scaling_factor[adapter_name] = module.in_features / qalora_group_size
+        else:
+            # No special scaling if dimensions don't align
+            module.qalora_scaling_factor[adapter_name] = 1.0
```

Comment:
Would this not lead to very different results than `qalora_scaling_factor = module.in_features / qalora_group_size`?

I wonder if perhaps it makes more sense to raise an error here and require that `module.in_features % qalora_group_size == 0`? I think this would simplify the code and when `module.in_features % qalora_group_size != 0` we just get standard LoRA?